File: Onera/Onera/Services/WebSearchService.swift
//
//  WebSearchService.swift
//  Onera
//
//  Web search service for injecting search context into LLM conversations.
//  All API calls are made client-side for E2EE compliance.
//

import Foundation

// MARK: - Search Result

struct WebSearchResult: Sendable {
    let title: String
    let url: String
    let snippet: String
    let content: String?
    let publishedDate: String?
    let score: Double?
}

// MARK: - Search Execution Result

struct WebSearchExecutionResult: Sendable {
    let provider: SearchProvider
    let query: String
    let results: [WebSearchResult]
    let timestamp: Date
}

// MARK: - Web Search Service Protocol

protocol WebSearchServiceProtocol: Sendable {
    func search(query: String, provider: SearchProvider, apiKey: String, maxResults: Int) async throws -> WebSearchExecutionResult
}

// MARK: - Web Search Service

final class WebSearchService: WebSearchServiceProtocol, Sendable {
    
    private let session: URLSession
    
    init(session: URLSession = .shared) {
        self.session = session
    }
    
    func search(query: String, provider: SearchProvider, apiKey: String, maxResults: Int = 5) async throws -> WebSearchExecutionResult {
        let results: [WebSearchResult]
        
        switch provider {
        case .tavily:
            results = try await searchTavily(query: query, apiKey: apiKey, maxResults: maxResults)
        case .brave:
            results = try await searchBrave(query: query, apiKey: apiKey, maxResults: maxResults)
        case .serper:
            results = try await searchSerper(query: query, apiKey: apiKey, maxResults: maxResults)
        case .exa:
            results = try await searchExa(query: query, apiKey: apiKey, maxResults: maxResults)
        }
        
        return WebSearchExecutionResult(
            provider: provider,
            query: query,
            results: results,
            timestamp: Date()
        )
    }
    
    // MARK: - Tavily
    
    private func searchTavily(query: String, apiKey: String, maxResults: Int) async throws -> [WebSearchResult] {
        let url = URL(string: "https://api.tavily.com/search")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "api_key": apiKey,
            "query": query,
            "max_results": maxResults,
            "include_answer": false,
            "include_raw_content": false
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw WebSearchError.providerError("Tavily search failed")
        }
        
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let results = json["results"] as? [[String: Any]] else {
            return []
        }
        
        return results.map { r in
            WebSearchResult(
                title: r["title"] as? String ?? "Untitled",
                url: r["url"] as? String ?? "",
                snippet: String((r["content"] as? String ?? "").prefix(300)),
                content: r["raw_content"] as? String,
                publishedDate: nil,
                score: r["score"] as? Double
            )
        }
    }
    
    // MARK: - Brave
    
    private func searchBrave(query: String, apiKey: String, maxResults: Int) async throws -> [WebSearchResult] {
        var components = URLComponents(string: "https://api.search.brave.com/res/v1/web/search")!
        components.queryItems = [
            URLQueryItem(name: "q", value: query),
            URLQueryItem(name: "count", value: String(maxResults))
        ]
        
        var request = URLRequest(url: components.url!)
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(apiKey, forHTTPHeaderField: "X-Subscription-Token")
        
        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw WebSearchError.providerError("Brave search failed")
        }
        
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let web = json["web"] as? [String: Any],
              let results = web["results"] as? [[String: Any]] else {
            return []
        }
        
        return results.map { r in
            WebSearchResult(
                title: r["title"] as? String ?? "Untitled",
                url: r["url"] as? String ?? "",
                snippet: r["description"] as? String ?? "",
                content: nil,
                publishedDate: r["age"] as? String,
                score: nil
            )
        }
    }
    
    // MARK: - Serper
    
    private func searchSerper(query: String, apiKey: String, maxResults: Int) async throws -> [WebSearchResult] {
        let url = URL(string: "https://google.serper.dev/search")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "X-API-KEY")
        
        let body: [String: Any] = [
            "q": query,
            "num": maxResults
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw WebSearchError.providerError("Serper search failed")
        }
        
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let results = json["organic"] as? [[String: Any]] else {
            return []
        }
        
        return results.map { r in
            WebSearchResult(
                title: r["title"] as? String ?? "Untitled",
                url: r["link"] as? String ?? "",
                snippet: r["snippet"] as? String ?? "",
                content: nil,
                publishedDate: r["date"] as? String,
                score: nil
            )
        }
    }
    
    // MARK: - Exa
    
    private func searchExa(query: String, apiKey: String, maxResults: Int) async throws -> [WebSearchResult] {
        let url = URL(string: "https://api.exa.ai/search")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        
        let body: [String: Any] = [
            "query": query,
            "numResults": maxResults,
            "useAutoprompt": true,
            "type": "neural"
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw WebSearchError.providerError("Exa search failed")
        }
        
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let results = json["results"] as? [[String: Any]] else {
            return []
        }
        
        return results.map { r in
            WebSearchResult(
                title: r["title"] as? String ?? "Untitled",
                url: r["url"] as? String ?? "",
                snippet: String((r["text"] as? String ?? "").prefix(300)),
                content: r["text"] as? String,
                publishedDate: r["publishedDate"] as? String,
                score: r["score"] as? Double
            )
        }
    }
    
    // MARK: - Format Results for Context
    
    /// Formats search results as context string to inject into the LLM prompt.
    /// Matches the web app's XML-like format.
    static func formatResultsForContext(results: [WebSearchResult], query: String) -> String {
        guard !results.isEmpty else { return "" }
        
        let formatted = results.enumerated().map { index, r in
            var result = "[\(index + 1)] \(r.title)\nURL: \(r.url)"
            if !r.snippet.isEmpty {
                result += "\n\(r.snippet)"
            }
            if let content = r.content, !content.isEmpty {
                let truncated = content.count > 500 ? String(content.prefix(500)) + "..." : content
                result += "\n\nContent:\n\(truncated)"
            }
            return result
        }.joined(separator: "\n\n---\n\n")
        
        return """
        <search_results query="\(query)">
        \(formatted)
        </search_results>
        """
    }
}

// MARK: - Web Search Error

enum WebSearchError: LocalizedError {
    case noProviderConfigured
    case noApiKey(String)
    case providerError(String)
    
    var errorDescription: String? {
        switch self {
        case .noProviderConfigured:
            return "No search provider configured"
        case .noApiKey(let provider):
            return "No API key configured for \(provider)"
        case .providerError(let message):
            return message
        }
    }
}

File: onera-android/app/src/main/java/chat/onera/mobile/data/remote/WebSearchService.kt
package chat.onera.mobile.data.remote

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.concurrent.TimeUnit

@Singleton
class WebSearchService @Inject constructor() {
    
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }
    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()
    
    enum class SearchProvider(val displayName: String) {
        TAVILY("Tavily"),
        BRAVE("Brave Search"),
        SERPER("Serper"),
        EXA("Exa")
    }
    
    @Serializable
    data class WebSearchResult(
        val title: String,
        val url: String,
        val snippet: String,
        val content: String? = null,
        val publishedDate: String? = null,
        val score: Double? = null
    )
    
    suspend fun search(
        query: String,
        provider: SearchProvider,
        apiKey: String,
        maxResults: Int = 5
    ): List<WebSearchResult> = withContext(Dispatchers.IO) {
        try {
            when (provider) {
                SearchProvider.TAVILY -> searchTavily(query, apiKey, maxResults)
                SearchProvider.BRAVE -> searchBrave(query, apiKey, maxResults)
                SearchProvider.SERPER -> searchSerper(query, apiKey, maxResults)
                SearchProvider.EXA -> searchExa(query, apiKey, maxResults)
            }
        } catch (e: Exception) {
            Timber.e(e, "Web search failed with ${provider.name}")
            emptyList()
        }
    }
    
    private fun searchTavily(query: String, apiKey: String, maxResults: Int): List<WebSearchResult> {
        val body = json.encodeToString(
            kotlinx.serialization.serializer<Map<String, String>>(),
            mapOf(
                "api_key" to apiKey,
                "query" to query,
                "max_results" to maxResults.toString(),
                "include_answer" to "false"
            )
        )
        val request = Request.Builder()
            .url("https://api.tavily.com/search")
            .post(body.toRequestBody("application/json".toMediaType()))
            .build()
        
        val response = httpClient.newCall(request).execute()
        val responseBody = response.body?.string() ?: return emptyList()
        return parseTavilyResponse(responseBody)
    }
    
    private fun searchBrave(query: String, apiKey: String, maxResults: Int): List<WebSearchResult> {
        val request = Request.Builder()
            .url("https://api.search.brave.com/res/v1/web/search?q=${java.net.URLEncoder.encode(query, "UTF-8")}&count=$maxResults")
            .get()
            .header("X-Subscription-Token", apiKey)
            .header("Accept", "application/json")
            .build()
        
        val response = httpClient.newCall(request).execute()
        val responseBody = response.body?.string() ?: return emptyList()
        return parseBraveResponse(responseBody)
    }
    
    private fun searchSerper(query: String, apiKey: String, maxResults: Int): List<WebSearchResult> {
        val body = json.encodeToString(
            kotlinx.serialization.serializer<Map<String, String>>(),
            mapOf("q" to query, "num" to maxResults.toString())
        )
        val request = Request.Builder()
            .url("https://google.serper.dev/search")
            .post(body.toRequestBody("application/json".toMediaType()))
            .header("X-API-KEY", apiKey)
            .build()
        
        val response = httpClient.newCall(request).execute()
        val responseBody = response.body?.string() ?: return emptyList()
        return parseSerperResponse(responseBody)
    }
    
    private fun searchExa(query: String, apiKey: String, maxResults: Int): List<WebSearchResult> {
        val body = json.encodeToString(
            kotlinx.serialization.serializer<Map<String, String>>(),
            mapOf(
                "query" to query,
                "num_results" to maxResults.toString(),
                "type" to "neural"
            )
        )
        val request = Request.Builder()
            .url("https://api.exa.ai/search")
            .post(body.toRequestBody("application/json".toMediaType()))
            .header("x-api-key", apiKey)
            .build()
        
        val response = httpClient.newCall(request).execute()
        val responseBody = response.body?.string() ?: return emptyList()
        return parseExaResponse(responseBody)
    }
    
    // Parse methods using lenient JSON parsing
    private fun parseTavilyResponse(body: String): List<WebSearchResult> {
        val data = json.decodeFromString<Map<String, JsonElement>>(body)
        val results = data["results"] as? JsonArray ?: return emptyList()
        return results.mapNotNull { element ->
            val obj = element as? JsonObject ?: return@mapNotNull null
            WebSearchResult(
                title = obj["title"]?.toString()?.trim('"') ?: "",
                url = obj["url"]?.toString()?.trim('"') ?: "",
                snippet = obj["content"]?.toString()?.trim('"') ?: "",
                score = obj["score"]?.toString()?.toDoubleOrNull()
            )
        }
    }
    
    private fun parseBraveResponse(body: String): List<WebSearchResult> {
        val data = json.decodeFromString<Map<String, JsonElement>>(body)
        val web = data["web"] as? JsonObject ?: return emptyList()
        val results = web["results"] as? JsonArray ?: return emptyList()
        return results.mapNotNull { element ->
            val obj = element as? JsonObject ?: return@mapNotNull null
            WebSearchResult(
                title = obj["title"]?.toString()?.trim('"') ?: "",
                url = obj["url"]?.toString()?.trim('"') ?: "",
                snippet = obj["description"]?.toString()?.trim('"') ?: ""
            )
        }
    }
    
    private fun parseSerperResponse(body: String): List<WebSearchResult> {
        val data = json.decodeFromString<Map<String, JsonElement>>(body)
        val organic = data["organic"] as? JsonArray ?: return emptyList()
        return organic.mapNotNull { element ->
            val obj = element as? JsonObject ?: return@mapNotNull null
            WebSearchResult(
                title = obj["title"]?.toString()?.trim('"') ?: "",
                url = obj["link"]?.toString()?.trim('"') ?: "",
                snippet = obj["snippet"]?.toString()?.trim('"') ?: ""
            )
        }
    }
    
    private fun parseExaResponse(body: String): List<WebSearchResult> {
        val data = json.decodeFromString<Map<String, JsonElement>>(body)
        val results = data["results"] as? JsonArray ?: return emptyList()
        return results.mapNotNull { element ->
            val obj = element as? JsonObject ?: return@mapNotNull null
            WebSearchResult(
                title = obj["title"]?.toString()?.trim('"') ?: "",
                url = obj["url"]?.toString()?.trim('"') ?: "",
                snippet = obj["text"]?.toString()?.trim('"')?.take(200) ?: ""
            )
        }
    }
    
    companion object {
        /**
         * Format search results as context to inject into the LLM prompt.
         */
        fun formatResultsForContext(query: String, results: List<WebSearchResult>): String {
            if (results.isEmpty()) return ""
            val entries = results.mapIndexed { i, r ->
                """<result index="${i + 1}">
<url>${r.url}</url>
<title>${r.title}</title>
<snippet>${r.snippet}</snippet>
</result>"""
            }.joinToString("\n")
            return """<search_results query="$query">
$entries
</search_results>"""
        }
    }
}
